_drafts/unit-test-mocking.md

```diff
@@ -0,0 +1,974 @@
+straight to your mailbox_
+
+{:.no_toc}
+
+## Table of Contents
+
+<!-- prettier-ignore -->
+* auto-gen TOC:
+{:toc}
+
+## Stubs, Mocks, and Fakes Review
+
+As a quick review, let's summarize the differences between **fakes**, **stubs**,
+and **mocks**.
+
+- **Fakes** will be a working implementation, but will usually substitute its
+  dependencies with something simpler and easier for a test environment.
+  Example: an in-memory key/value store vs a NOR-flash backed Key/Value store.
+- **Stubs** will be an implementation that returns canned values, generally
+  always returning valid or invalid values.
+- **Mocks** are an implementation that is controlled by the unit test. They can
+  be pre-programmed with return values, check values of arguments, and help
+  verify that functions are called.
+
+Each one serves a purpose in unit testing. Although each type might not be
+utilized in every unit test you write, I can guarantee you that there's a place
+for them within the suite of tests for your firmware.
+
+## Mocks in Detail
+
+A mock is a simulated function or module that mimics the behavior of a real
+implementation but is fully controlled by the unit test itself. The programmer
+can validate how many times a mock is is called and verify the value of all
+arguments passed into the mock. The programmer also dictates what actions the
+mock takes, from calling other fakes or stubs, to returning particular return
+values, or modifying certain output parameters.
+
+Here are some common use cases for mocks that we'll cover in this post:
+
+- A malloc implementation that can be pre-programmed with return values (return
+  real buffers vs NULL).
+- A mock flash driver that returns error codes and forces different paths in a
+  higher level module.
+- A Bluetooth socket implementation that is fed artfully crafted packed data to
+  instrument protocols.
+
+Mocks are especially useful when testing failure paths that would likely never
+happen, or be disastrous if they did happen. For example, a mock would be a good
+choice if you are writing a vehicle software module that behaves differently if
+the airbags have triggered or not. You _really_ would not want to have to test
+this in QA every time you make a change.
+
+Since mocks generally do not include any other dependencies, mocks are also the
+best tool in the toolbox for isolating and testing every facet of the module
+under test (strive for that 100% code coverage).
+
+However, this strength is also a weakness. Using mocks everywhere becomes
+cumbersome and verbose, given that every call, parameter, and return value has
+to be explicitly defined or ignored. Everything in testing is a trade-off, and
+choosing when to use mocks is no exception.
+
+### When to Use Mocks
+
+You should considering using a mock:
+
+- When you only want to include single `.c` or `.cpp` source file under test.
+- When you feel you need something more control than a **stub** but less than a
+  **fake**.
```

Comment:
When you need more than a stub, but don't want to go through the trouble of modeling subsystem behavior perfectly in a fake